### PR TITLE
genpartial: refactor to type traversal

### LINKED ISSUE #1129: `genpartial` sometimes skips certain fileds (inline structs?)
As discovered in https://github.com/redpanda-data/helm-charts/pull/1121, `genpartial` can sometimes skip over anonymous structs. The issue has been worked around so this isn't burning but it would be nice to root cause and then fix.

JIRA Link: [K8S-132](https://redpandadata.atlassian.net/browse/K8S-132)

[K8S-132]: https://redpandadata.atlassian.net/browse/K8S-132?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### PR DESCRIPTION
**Pull request by chrisseto**
This commit refactors `genpartial` to instead traverse types via the go type system instead of traversing the AST. This strategy fixes a handful of bugs that previously went unnoticed and otherwise makes the package a bit more approachable.

Fixes #1129

### CODE REVIEW

**Comment by RafalKorepta** [comment]
```
-: # github.com/redpanda-data/helm-charts/charts/redpanda_test [github.com/redpanda-data/helm-charts/charts/redpanda.test]
charts/redpanda/chart_template_test.go:165:24: undefined: redpanda.PartialTunableConfig
charts/redpanda/chart_test.go:32:20: undefined: redpanda.PartialNodeConfig
charts/redpanda/chart_test.go:40:12: cannot use &redpanda.PartialTiered{…} (value of type *redpanda.PartialTiered) as *redpanda.Tiered value in struct literal
charts/redpanda/chart_test.go:41:23: undefined: redpanda.PartialTieredStorageConfig
charts/redpanda/chart_test.go:76:20: undefined: redpanda.PartialNodeConfig
charts/redpanda/chart_test.go:84:12: cannot use &redpanda.PartialTiered{…} (value of type *redpanda.PartialTiered) as *redpanda.Tiered value in struct literal
charts/redpanda/chart_test.go:86:17: cannot use &redpanda.PartialSecretRef{…} (value of type *redpanda.PartialSecretRef) as *redpanda.SecretRef value in struct literal
charts/redpanda/chart_test.go:87:17: cannot use &redpanda.PartialSecretRef{…} (value of type *redpanda.PartialSecretRef) as *redpanda.SecretRef value in struct literal
charts/redpanda/chart_test.go:89:23: undefined: redpanda.PartialTieredStorageConfig
charts/redpanda/chart_test.go:117:22: undefined: redpanda.PartialNodeConfig
```
There are errors around `chart_test.go` reported by our CI
https://github.com/redpanda-data/helm-charts/actions/runs/9650814469/job/26617281354?pr=1354#step:5:414

**Comment by chrisseto** [comment]
I was trying to be too clever with deduping types. Removed the cleverness and finally figured out how to make AST printing deterministic!

File: cmd/genpartial/main.go [hunk 1]
```diff
@@ -66,151 +67,225 @@ func main() {
 	}
 }
 
-func GeneratePartial(pkg *packages.Package, structName string, out io.Writer) error {
-	root := pkg.Types.Scope().Lookup(structName)
+type Generator struct {
+	pkg     *packages.Package
+	cache   map[types.Type]ast.Expr
+	imports map[string]types.PkgName
```
Comment (RafalKorepta) [inline cmd/genpartial/main.go:73]: This filed is unused per staticcheck 

```
task: [ci:lint] staticcheck ./...
verifying file: charts/redpanda/ci/32-statefulset-podspec-novalues.yaml
verifying file: charts/redpanda/ci/33-advertised-ports-values.yaml
verifying file: charts/redpanda/ci/33-pod-selector-lables-novalues.yaml
verifying file: charts/redpanda/ci/34-security-contexts-novalues.yaml
verifying file: charts/redpanda/ci/34-statefulset-sidecars-novalues.yaml
verifying file: charts/redpanda/ci/35-connectors-novalues.yaml
verifying file: charts/redpanda/ci/36-single-external-address-with-template-domain-novalues.yaml
verifying file: charts/redpanda/ci/37-internal-service-changed-name-and-annotations-novalues.yaml
verifying file: charts/redpanda/ci/38-post-install-upgrade-merges-novalues.yaml
verifying file: charts/redpanda/ci/38-post-install-upgrade-no-overrides-novalues.yaml
verifying file: charts/redpanda/ci/96-audit-logging-values.yaml.tpl
verifying file: charts/redpanda/ci/97-license-key-values.yaml.tpl
verifying file: charts/redpanda/ci/98-license-secret-values.yaml
verifying file: charts/redpanda/ci/99-none-existent-config-options-with-empty-values.yaml
cmd/genpartial/main.go:73:2: field imports is unused (U1000)
```

https://github.com/redpanda-data/helm-charts/actions/runs/9666564401/job/26666371780?pr=1354#step:5:415

Comment (chrisseto) [reply cmd/genpartial/main.go]: Fixed!

**Review by RafalKorepta — APPROVED** [review]

**Pull request merged**